openassessment/xblock/grade_mixin.py

```diff
@@ -584,3 +590,101 @@ def _assessment_grade_context(self, assessment):
                     part['option']['label'] = option_labels.get(option_label_key, part['option']['name'])
 
         return assessment
+
+    def _get_assessment_type(self, workflow):
+        # assessment_step would always have staff in it, so be careful while
+        # making decisions upon it
+        _assessment_type = None
+        score = workflow['score']
+        complete = score is not None
+
+        is_self_assessment = "self-assessment" in self.assessment_steps
+        is_peer_assessment = "peer-assessment" in self.assessment_steps
+        is_staff_assessment = "staff-assessment" in self.assessment_steps
+
+        if is_staff_assessment:
+            _assessment_type = "staff_defined"
+        else:
+            # Staff assessment is not a requirement but it can be overrided
+
+            # Check for staff override
+            if complete:
+                # If an assessment is done by staff it always have related annotation
+                _assessment_type = (
+                    "staff_defined"
+                    if "staff_defined" in [
+                        annotation['annotation_type'] for annotation in score["annotations"]
+                    ]
+                    else None
+                )
+
+            # if it is not staff assessment and it also isn't overrided by staff
+            if _assessment_type is None:
+                try:
+                    if not is_peer_assessment:
+                        raise ValueError()
+                    peer_index = workflow["assessment_score_priority"].index("peer")
+                except ValueError:
+                    peer_index = float("inf")
+
+                try:
+                    if not is_self_assessment:
+                        raise ValueError()
+                    self_index = workflow["assessment_score_priority"].index("self")
+                except ValueError:
+                    self_index = float("inf")
+
+                if min(self_index, peer_index) == peer_index:
+                    _assessment_type = "peer_defined"
+                else:
+                    _assessment_type = "self_defined"
```

Comment:
Taking into account my other suggestions, and changing from `peer_defined` to just `peer` or `self` (without the `_defined`), the code can be boiled down to:
```python
for score_priority in workflow["assessment_score_priority"]:
    # Skip staff
    if score_priority == "staff":
        continue

    # Return first assessment type (the one with the higest priority is the first).                 
    return score_priority
```

The entire function can be globally optimized to get a smaller footprint like this:
```python
    def _get_assessment_type(self, workflow):
        """
        Some docstring here :)
        """
        score = workflow['score']
        complete = score is not None

        if "staff-assessment" in self.assessment_steps:
            return "staff"

        # Edge case: staff overrides the grade.
        # If a score is overriden by staff, it'll always have an
        # attached annotation type with the `staff_defined` value,
        # so we look for that in this problem's annotation and
        # return staff if it's found.
        grade_annotation_types = [annotation['annotation_type'] for annotation in score["annotations"]]
        if complete and "staff_defined" in grade_annotation_types:
            return "staff"

        # For other cases, we just need to figure out the
        # priority of each (either peer or self).
        # Just loop over the values and return the first one
        # after staff.
        for _assessment_type in workflow["assessment_score_priority"]:
            # assessment_step would always have staff in it, so skip it 
            # while checking the priority here.
            if _assessment_type == "staff":
                continue

            return _assessment_type
```
It's simpler than the current approach and is more human readable.
Take care to adjust `_get_score_explanation` to remove the `_defined` suffix.